tests/programs/custom/zalrsc/zalrsc.c

```diff
@@ -0,0 +1,216 @@
+
+  } else {
+    return EXIT_FAILURE;
+  }
+}
+
+int test_LR_SC_sequence_success(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to succeed.\n\n");
+
+  volatile uint32_t mem_addr = 0;
+  volatile uint32_t is_failure = 0;
+  volatile uint32_t random = 0;
+
+  mem_addr = (uint32_t)&mem_addr;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)     : "r" (mem_addr));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure) : "r" (random), "r" (mem_addr));
+
+  if (is_failure) {
+    printf("Expected sequence to succeed. However, it did not.\n");
+    return 1;
+  }
+
+  return 0;
+}
+
+int test_LR_SC_sequence_failure_address_mismatch(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to fail due to SC.W address mismatch.\n\n");
+
+  volatile uint32_t mem_addr_LRW = 0;
+  volatile uint32_t mem_addr_SCW = 0;
+  volatile uint32_t is_failure = 0;
+  volatile uint32_t random = 0;
+
+  mem_addr_LRW = (uint32_t)&mem_addr_LRW;
+  mem_addr_SCW = (uint32_t)&mem_addr_SCW;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)      : "r" (mem_addr_LRW));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure)  : "r" (random), "r" (mem_addr_SCW));
+
+  if (!is_failure) {
+    printf("Expected sequence to fail. However, it did not.\n");
+    return 1;
+  }
+
+  return 0;
+}
+
+int test_LR_SC_sequence_failure_additional_SC_access(void)
+{
+  printf("\n\nTest LR.W SC.W sequence constructed to fail due to additional SC.W access, before SC.W access the address LR.W original reserved.\n\n");
+
+  volatile uint32_t mem_addr;
+  volatile uint32_t mem_addr_additional_SCW;
+  volatile uint32_t is_failure_additional_SCW;
+  volatile uint32_t is_failure;
+  volatile uint32_t random;
+
+  mem_addr                = (uint32_t)&mem_addr;
+  mem_addr_additional_SCW = (uint32_t)&mem_addr_additional_SCW;
+
+  __asm__ volatile("lr.w %0, (%1)"     : "=r" (random)                    : "r" (mem_addr));
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure_additional_SCW) : "r" (random), "r" (mem_addr_additional_SCW)); //TODO: krdosvik, not 100% sure if this should fail?
+  __asm__ volatile("sc.w %0, %1, (%2)" : "=r" (is_failure)                : "r" (random), "r" (mem_addr));
+
+  if (!is_failure_additional_SCW) {
+    printf("Expected SCW instruction to fail. However, it did not.\n");
+    return 1;
+  } else if (!is_failure) {
+    printf("Expected sequence to fail. However, it did not.\n");
+    return 1;
+  }
+
+  return 0;
+}
+
+int test_LRW_misaligned_access(void)
+{
+  printf("\n\nTest that a misaligned LR.W instruction fails.\n\n");
+
+  volatile uint32_t mem_addr_misaligned;
+  volatile uint32_t random;
+
+  mem_addr_misaligned = ((uint32_t)&mem_addr_misaligned) +1;
+  trap_handler_beh = MISALIGNED_LRW;
+
+  __asm__ volatile("lr.w %[random], (%[mem_addr_misaligned])" : [random]"=r" (random) : [mem_addr_misaligned]"r" (mem_addr_misaligned));
+
+  //(Execute exception handler)
+
+  if (num_misaligned_LRW_trapped != 1) {
```

Comment:
This exact number depends on the order the tests were run in? (Or, if you run that test twice it fails?)

It seems vulnerable to future breakage.
But since we don't expect to do any amendments that could cause such breakage, I think it is fine as it is.

Reply:
I dont think there is a order dependent on the tests, but it is indeed a number dependent on the tests, so running the test_LRW_misaligned_access or test_SCW_misaligned_access twice or more will make them fail. I agree that it is unlikely that we will run the tests several times, but as you say it is indeed a vulnerability which can be easily fixed by setting the values to zero when exiting, so will do that. Thanks!

Reply:
fixed